Handle multiple results per document in a sync batch MERGEOK

bratseth opened the pull request:
@vekterli please review

Commit:
Remove deprecated and unused method

arnej27959 committed:
use existing container cluster if possible

gjoranv committed:
Remove unused configs and their usage.

gjoranv committed:
Remove deprecated and unused TimeoutCollector/TimeoutRate

gjoranv committed:
Don't install removed def files.

gjoranv committed:
Remove config param noSearchBackendsImpliesOutOfService

gjoranv committed:
Fix errors from rebasing to master.

bjorncs committed:
Remove spooler from config-model

bjorncs committed:
Remove soLingerTime from connector config

bjorncs committed:
Remove old ssl connector config

bjorncs committed:
Move BouncyCastle preinstall declaration to vespajlib

bjorncs committed:
Fix todos in Cookie

bjorncs committed:
Remove deprecated escaping

gjoranv committed:
Re-add file removed by accident during rebase to master.

gjoranv committed:
Remove unused ctor in InnerNodeVector

arnej27959 committed:
remove leftover debug prints

gjoranv committed:
Remove deprecated DoubleFormatter with tests.

gjoranv committed:
Remove deprecated DoubleParser.

gjoranv committed:
Remove deprecated LanguageHacks.

gjoranv committed:
Remove deprecated metrics (*MemoryBytes)

gjoranv committed:
Require JDK 9+ for tenant-base.

bjorncs committed:
Move BouncyCastle preinstall declaration to container-disc

bjorncs committed:
Disable TLS RSA ciphers that do not support forward secrecy

This will essentially remove the temporary workaround introduced with
the Jetty 9.4.12 upgrade that was done recently. JDisc will with this
change only enable certificates marked as grade A by ssllabs.com

Commit:
Remove unused config settings.

Commit:
Track removal of fileHandler.rotateSchema from access-log.def

gjoranv committed:
Remove deprecated and unused BindingMatch ctor.

Commit:
Remove further traces of rotationScheme.

gjoranv committed:
Remove unused qr-logging.def

Commit:
Remove deprecated settings from proton.def.

bjorncs committed:
Remove com.yahoo.jdisc.application and com.yahoo.jdisc.client as public api

bjorncs committed:
Remove jdisc-logfmt

bjorncs committed:
Remove TODO that is currently not feasible to solve

Commit:
Remove deprecated memorytouse setting in stor-server.def.

Commit:
Remove deprecated iterators_per_bucket setting from stor-visitor.def.

Commit:
Remove workarounds added during transition to multiple bucket spaces.

geirst committed:
Remove all legacy proton metrics (that were deprecated on Vespa 6).

geirst committed:
Remove AttributeMetricsCollection and DocumentDBMetricsCollection as they are no longer needed.

geirst committed:
Remove metrics for persistence provider that were deprecated and broken (not thread safe) on Vespa 6.

geirst committed:
Remove deprecated proton tuning settings.

bjorncs committed:
Remove deprecated Manhattan metrics

bjorncs committed:
Remove AccessLogEntry.setURI/getURI

bjorncs committed:
Remove TODO comment that was leftover from previous refactoring

gjoranv committed:
Upgrade Felix to 6.0.1

- Exclude artifact that is not needed nor provided runtime.

geirst committed:
Remove deprecated APIs.

Commit:
Only allow document api protocol version 6.221 (Java)

gjoranv committed:
Ignore test that fails with Felix 6.0.

- The jdisc_core "bundle" now exports java.* packages, which
  only the felix.framework system bundle is allowed to.

Commit:
Merge documentapi protocol routable factories (Java).

Commit:
Only allow document api protocol version 6.221 (C++)

Commit:
Merge documentapi protocol routable factories (C++).

Commit:
Remove cross language test files for old document api version.

gjoranv committed:
Retrieve the ServiceReference via the service's own bundle.

- Going via the framework/bundleContext does not work, because
  it's class is loaded by the AppClassLoader, which will then be
  used to lookup the service class. Since this is a unit test,
  the class will be found on the class path but, since it's a
  different class than the service class in the bundle, a class
  cast check will fail, and the returned service ref will be null.
- Note that this worked on Felix <6.0 because an exception was
  thrown (because the framework bundle does not export our service
  class), which led to the class cast check being skipped
  altogether.
  See felix/framework:ExtensionManager.getClassByDelegation
  (commit 3c147cb58a5)

gjoranv committed:
Bootstrap FelixParams at runtime.

- The set of system packages is not constant (nor is their order)
  between JVM invocations.

gjoranv committed:
Add TODO to get rid of mockito-all dependency

gjoranv committed:
Extract private helper for finding the set of exported packages.

gjoranv committed:
Use JDK 11

gjoranv committed:
Upgrade maven-plugin-tools artifacts, and use a version property

bjorncs committed:
Revert back to original logcontrol directory for jdisc log

gjoranv committed:
Remove deployState parameter, forgotten during rebase to master.

- Was used for spooler only.

bjorncs committed:
Generate xsd+rng in target directory

bjorncs committed:
Don't build twice

arnej27959 committed:
remove "defidx" and related classes

gjoranv committed:
Upgrade guava to 20.

- guava 21.0 removes apis used by curator/zookeeper (and hadoop).
  We could embed the older version in zkfacade bundle, but
  it requires more pom maintenance. Also, upgrading beyond 23 will
  require changes in OsgiUtil, causing a non-resolvable conflict
  in orchestrator unit tests which will then require old and new
  guava at the same time. (This can probably be solved by moving
  application based unit tests to container-integration-test.)

Commit:
Remove legacy form

gjoranv committed:
Remove Vespa 7 TODO that is unfeasible.

gjoranv committed:
Remove deprecated apis in linguistics.

bjorncs committed:
Make Json default output format for vespaget and vespavisit

- Also introduce --xmloutput parameter

Commit:
Inline test files

gjoranv committed:
Remove deprecated apis in docproc.

gjoranv committed:
Inline init method.

- Failed on assignment to final field.

geirst committed:
Stop using legacy proton metrics that have been removed on Vespa 7.

gjoranv committed:
Open necessary packages to the unnamed module.

- To avoid warnings due to Felix accessing non-public apis with
  reflection.

gjoranv committed:
Open necessary packages to the unnamed module

.. to avoid warnings due to Felix accessing non-public apis with
  reflection.
+ Show all "Illegal reflective access" warnings.

gjoranv committed:
Service name is 'qrserver' again, since PR #7332

Commit:
6-SNAPSHOT -> 7-SNAPSHOT

arnej27959 committed:
remove container.legacy-emulation config

Commit:
A few more updates to 7-SNAPSHOT.

Commit:
Update VERSION to 7.0.0.

gjoranv committed:
CMS is deprecated in JDK 9 and will be removed in a later release

- See http://openjdk.java.net/jeps/291

gjoranv committed:
Open necessary packages to the unnamed module

- for configserver and yet another standalone start script

gjoranv committed:
Remove built-in support for JAXBContextFactory.

Commit:
6-SNAPSHOT -> 7-SNAPSHOT.

arnej27959 committed:
remove "attribute: prefetch" code

arnej27959 committed:
remove "attribute: prefetch" in tests

arnej27959 committed:
skip fields which are empty or only have the default value

arnej27959 committed:
Revert "skip fields which are empty or only have the default value"

This reverts commit e7c9bc3b5d85c2eafef8cd06017304fe0ff3dc47.

arnej27959 committed:
skip fields which are empty or only have the default value

arnej27959 committed:
use getUndefined from searchcommon undefinedvalues

gjoranv committed:
Remove done TODO and add comments on why we need jaxb deps.

baldersheim committed:
remove deprecated methods and switch to hashmap instead of list.

arnej27959 committed:
remove "header" and "body" keywords

gjoranv committed:
Use G1GC by default on Vespa 7 (JDK 9+).

Commit:
Make stemming.best default

arnej27959 committed:
stop using "header" and "body" keywords

arnej27959 committed:
no more body fields

arnej27959 committed:
remove "header" from embedded SD

arnej27959 committed:
default to JSONAccessLog

arnej27959 committed:
Revert "remove "header" and "body" keywords"

This reverts commit e24c1df653d6470089f05786d74e729bb8f82990.

Commit:
Make SimpleLinguistics simple again

- Remove SimpleLinguistics config and optional use of Optimaize
- Add Optimaize to OpennlpLinguistics; on by default and config to disable

Commit:
Follow API change

Commit:
Follow API change

Commit:
Change default timeout from 5000 to 500 ms

Commit:
Turn softtimeout on by default

Commit:
Make OpenNlpLinguistics default

Commit:
Rename constant

Commit:
Switch default data format for vespa http client to json.

gjoranv committed:
Fixes after rebase.

- Update gc options test
- Suppress deprecation warnings temporarily

Commit:
Remove deprecated code

Commit:
Rename DefaultRenderer to XmlRenderer

Commit:
Remove HTTP searchers

geirst committed:
Stop renaming metrics in VespaMetricSet.

arnej27959 committed:
remove unused classes (again)

gjoranv committed:
Update method name. (rebase issue?)

Commit:
Remove support for vespa http client protocol v2.  Always use protocol v3.

geirst committed:
Use G1GC as default on Vespa 7 (fixup after rebasing).

gjoranv committed:
Implement new (static) method in Reader.

+ Improve test to print out unmasked methods.

gjoranv committed:
Upgrade asm to 7 and use Opcodes.ASM7

gjoranv committed:
Upgrade maven-shade-plugin for JDK 11

- Uses asm 7

gjoranv committed:
Improve integration test

- Add nested classes
- Put back commented assert and remove stale comment
+ Add comment to replace jrt.jar. It's compile scoped and had a
  much smaller dependency tree when it was added compared to now.
  Now, it adds loads of Import-Packages, increasing the risk of
  invalidating the integration test that verifies that the
  required imports have been added to the manifest.

Commit:
Remove old feedapi

Commit:
Remove template support

Commit:
Remove wrong TODO - this is needed for vespafeeder

Commit:
Do Vespa 7 TODOs

Commit:
Remove wrong TODO

Commit:
Mark Vespa 8

Commit:
Remove unused imports

gjoranv committed:
Clarify that deprecation will be removed for Vespa 8

gjoranv committed:
Don't install removed def file

gjoranv committed:
Remove binary-cache.def

arnej27959 committed:
remove deprecated methods

Commit:
Remove deprecated method (again)

arnej27959 committed:
remove or disable --headersonly flag

arnej27959 committed:
remove streaming.headersonly option

arnej27959 committed:
ignore headersonly flag

* methods taking the flag are now deprecated, added new variants not taking the flag

Commit:
Remove deprecated method

arnej27959 committed:
remove unused "headersonly" functionality

arnej27959 committed:
fix test to not use (removed) "-e" option

Commit:
Remove internal http feed handler protocol version 2.

Commit:
Remove deprecated methods

Commit:
Do TODOs

Commit:
Remove traces of protocol version 2 from internal http feeder.

gjoranv committed:
Java 11: Update ExportPackages:

- Packages exported from Felix:
  * com.sun.security.*
  * com.sun.management
  * sun.misc
  * org.w3c.dom.*

- Packages hidden in modules:
  * sun.net.util

- Packages that don't exist in JDK 11:
  * sun.security.krb5

gjoranv committed:
Bugfix: make NestedStaticClass static

gjoranv committed:
Use JDK 11 docker image in Travis

Commit:
Don't set up search implicitly

Commit:
Remove dead code

Commit:
Optimize imports

Commit:
Require methods with timeout to be implemented

Commit:
Remove scoreboard

Commit:
Register page templates xml renderer

Commit:
Remove coverageReport parameter - always on

baldersheim committed:
Order is changed.

baldersheim committed:
This should be better order.

arnej27959 committed:
nobody uses JSVC aka commons.daemon, remove support for it

gjoranv committed:
Partial cleanup of metric-defaults config.

arnej27959 committed:
just ignore header/body keywords

arnej27959 committed:
GC dead code

gjoranv committed:
Do not provide apache httpclient and httpcore.

gjoranv committed:
Move TODO to Vespa 8

gjoranv committed:
Remove httpclient dependency.

- Should probably have been cleaned up when code was moved to
  vespa-athenz.

gjoranv committed:
Remove metric-defaults.def

- Contained only one enum with one value.

arnej27959 committed:
update or add comments around deprecated methods

arnej27959 committed:
we expect the "default" summary to fetch fields from disk

gjoranv committed:
Don't install removed def file

gjoranv committed:
Remove @PublicApi from package c.y.document.select

Commit:
Deprecation cleanup

arnej27959 committed:
add class with complete serializer for 6.x format

arnej27959 committed:
wire in new Serializer6

arnej27959 committed:
actually use Serializer6

arnej27959 committed:
update test now that "default" is always "from-disk"

gjoranv committed:
Generate html5 javadoc

arnej27959 committed:
just confusing to have Deserializer extend Serializer

arnej27959 committed:
add class with complete deserializer for the 6.x format

arnej27959 committed:
remove split header/body buffers support

arnej27959 committed:
wire in Deserializer6

arnej27959 committed:
test Deserializer6 more

arnej27959 committed:
split header/body is very deprecated

arnej27959 committed:
remove split header/body methods

gjoranv committed:
Disable offline link detection

Error message:
Error fetching link: /Users/gv/git/vespa/testutil/target/site/apidocs/package-list. Ignored it.

arnej27959 committed:
avoid using a dynamic documentlist just to hold a single operation

arnej27959 committed:
remove unused class and imports

arnej27959 committed:
remove unused imports

arnej27959 committed:
remove unused classes

arnej27959 committed:
remove "headerOnly" serialization

arnej27959 committed:
track API change

arnej27959 committed:
minor refactoring of document serialization

arnej27959 committed:
remove misleading usage of DocumentSerializerFactory

arnej27959 committed:
generate "serializeupdatejava.dat" in old format for now

gjoranv committed:
Open package to avoid warning during jdisc_core compile.

gjoranv committed:
Open java.base/java.io

.. to allow com.yahoo.io.NativeIO to use reflective access on
   java.io.FileDescriptor.fd

gjoranv committed:
G1GC is default on 7

gjoranv committed:
Remove unused ComponentClass

gjoranv committed:
Add comments to AbstractComponent constructors.

gjoranv committed:
Clarify that these classes do not have to be implemented for DI.

geirst committed:
Remove remains of RISE.

geirst committed:
Temporarily ignore unstable test on Vespa 7.

geirst committed:
Remove unused settings in proton.def.

summary.log.compact2activefile is always false.
summary.compact2buckets is always true.

arnej27959 committed:
remove unused code

arnej27959 committed:
remove unused code

Commit:
Provide enough time to run on factory

arnej27959 committed:
disallow absolute paths

arnej27959 committed:
bump version in TODOs into the future

arnej27959 committed:
throw away pre-6.0 code

arnej27959 committed:
disallow old top-level "cluster", hope nobody uses it
update todo for old top-level tag "config", used by several unit tests.

geirst committed:
Remove unused search.io setting.

gjoranv committed:
Add version-less exports for some javax packages

.. to compensate for the fact that 6 exported these packages with
   both the artifact version and "0.0.0.1_008_JavaSE". The latter
   takes precedence because it was listed first in the manifest,
   such that user bundles import these package with range [0,1).

gjoranv committed:
Add space after commas.

- It's common practice, and makes exportPackages.properties
  easier to split.

gjoranv committed:
Set 7-SNAPSHOT for new modules.

bjorncs committed:
Add back method that was intended to be removed in Vespa 7

bjorncs committed:
Ignore .rng and .xsd files to reduce git noise

bjorncs committed:
Makefile is committed - do not ignore

arnej27959 committed:
provide defaults, making it easier to mock up an implementation

gjoranv committed:
Do not provide commons-codec at build time.

- We effectively stopped providing it runtime since
  https://github.com/vespa-engine/vespa/pull/7696

arnej27959 committed:
catch up with Vespa 7 work

* use vespaversion as default for test_framework_version
* target Java 11

geirst committed:
Remove values that are removed in config def file on Vespa 7.

geirst committed:
Remove values that are removed in config def file on Vespa 7.

gjoranv committed:
Set version to 7-SNAPSHOT for new module.

gjoranv committed:
JDK 11 updates.

- Use Opcodes.ASM7
- Inherit maven-plugin-api version from parent (currently 3.6.0)
- Use enforcer-plugin instead of Maven 2 prerequisites
- Require maven 3.5 as for the rest of Vespa
- Inherit asm version from parent to stay updated

Commit:
Nonfunctional changes only

Commit:
Testing of dynamic summary rendering

arnej27959 committed:
add code to insert position attributes directly

arnej27959 committed:
use IntegerContent to retrieve positions

Commit:
Don't expect uri field

Commit:
Remove Query.getSessionId(boolean)

Commit:
Remove Sever

Commit:
Remove deprecated methods

Commit:
Remove deprecated methods

Commit:
Revert "Bratseth/search container deprecations"

Commit:
Revert "Merge pull request #7974 from vespa-engine/revert-7973-bratseth/search-container-deprecations"

This reverts commit 5767288d80e85940e675f7dd157f2b6c1d2ceeb5, reversing
changes made to 37de7ac33454363489156e06597bffb3422756dc.

gjoranv committed:
Set 7-SNAPSHOT for new module, and remove redundant 'name'.

Commit:
Remove ymum from grouping

Commit:
Remove deprecated duplicate Version class

gjoranv committed:
Remove removed ctor in InnerNodeVector from abi spec.

gjoranv committed:
Update abi specs for 7.

geirst committed:
Render single value position as object to match how top-level position fields are rendered.

baldersheim committed:
Follow api change

Commit:
Repair serialization of boolean fields.

Commit:
Handle multiple results per document in a sync batch

bratseth commented:
There's some cleanup in the way, but the point of this is to index results in a sync batch by operation id rather than document id. This necessitates allowing the operation id to be decided by the caller.

Commit:
Set -XX:ActiveProcessorCount option if set in config
Use VESPA_CONTAINER_JVMARGS if set

vekterli reviewed (commented):
I have some questions related to ABI compatibility and how explicit operation IDs will be used in practice that I would like to resolve first. Otherwise LGTM.

bratseth commented on vespa-http-client/src/main/java/com/yahoo/vespa/http/client/FeedClient.java:
```diff
@@ -39,8 +41,21 @@
+     * @param documentData the document data as JSON or XML (as specified when using the factory to create the API)
+     * @param context a context object which will be accessible in the result of the callback, or null if none
+     */
+    void stream(String documentId, String operationId, CharSequence documentData, Object context);
```
Yes, it would break mocking. I decided against doing something slightly uglier to avoid that since this goes into the second 7 release and I doubt anybody is mocking this in any case.

bratseth commented on vespa-http-client/src/test/java/com/yahoo/vespa/http/client/core/operationProcessor/OperationProcessorTest.java:
```diff
@@ -27,16 +27,15 @@
+    final Document doc1 = new Document("doc:a:b", null, "data doc 1", null);
+    final Document doc1b = new Document("doc:a:b", null, "data doc 1b", null);
+    final Document doc2 = new Document("doc:a:b2", null, "data doc 2", null);
+    final Document doc3 = new Document("doc:a:b3", null, "data doc 3", null);
```
Yes, added.

bratseth commented on vespa-http-client/src/main/java/com/yahoo/vespa/http/client/core/api/FeedClientImpl.java:
```diff
@@ -45,17 +45,12 @@ public FeedClientImpl(SessionParams sessionParams,
         charsetEncoder.onUnmappableCharacter(CodingErrorAction.REPORT);
 
-        Document document = new Document(documentId, documentData, context);
+        Document document = new Document(documentId, operationId, documentData, context);
```
We don't want to specify the same operation id for multiple documents!
The problem is that the sender may need to know about the operation id before receiving the callback for the operation (to determine if it issued the operation), and if the operation id is not known before after issuing the operation we enter a race between the callback and the sender preparing to receive it. 

The simple way to avoid this is to let the sender decide it ... with the drawback that we introduce a new shiny way for it to shoot self in foot. I'll add some doc on this.

vekterli replied:
Got it, was a bit confused as to the relationship between an operation and a document ID in this context.

bjorncs committed:
Fix accidental import of java.sql.Date

Commit:
Adjust java dependencies and path to java libraries.

Commit:
Deployment spec majorVersion take precedence over Application value

Overriding the major version set in the deployment spec does not work
because config servers will filter out major versions based on the
deployment spec, and it is not useful because we should respect
applications explicit choices when specified.

baldersheim committed:
No more docprocservice.

Commit:
Don't expect docprocservice as service type

baldersheim committed:
remove test that does not make that much sense.

Commit:
Revert "Default to system version when deploying to dev"

This reverts commit 416ae7ee0bc8431eca0a4de5c014c232bd786f06.

Commit:
Use major from deployment.xml when deploying to dev/perf

Commit:
Copr does not allow using the alternatives command. Setting JAVA_HOME is enough for mvn to pick up JDK 11.

Commit:
We execute the java binary explicitly when compiling. Need to adjust PATH. Also parallell compilation with mvn works now.

bjorncs committed:
Build master branch

Commit:
Use same reservation expiry in all zones

There is no difference in how we build models in dev and other zones
anymore, and there are customers with large apps that need more than 10 minutes

toregge commented:
Missing rebase before merging ?

bratseth commented:
Missing pull before rebasing .. then rebasing again ... apparently lead to this. The diff to master was correct so I didn't bother finding the right commits and doing it over.

toregge commented:
This broke git history (many commits appear multiple times in history).

bratseth commented:
Alright I'll revert and submit it in a different PR then.

toregge commented:
That won't fix history.

bratseth commented:
Yes, it doesn't help.
